config/config.go

```diff
@@ -722,6 +725,25 @@ func expandEnv(s string) string {
 	})
 }
 
+func unmarshaler(factory component.Factory) component.CustomUnmarshaler {
+	if du, ok := factory.(deprecatedUnmarshaler); ok {
```

Comment:
Shouldn't you check for non-deprecated interface first?
i.e. if factory implements both new and old interfaces, we should use the "new" one?

Reply:
Done.